File: src/puzzle.rs
use std::{
    collections::HashSet,
    iter::Sum,
    ops::{Add, AddAssign},
};

use bevy::{
    math::{UVec2, Vec2}, // TODO: Depend directly on glam if we split crates?
    prelude::Component,
};

use itertools::Itertools;

/// A `Puzzle` grid, with no associated solution
#[derive(Debug, Clone, Component)] // TODO: Not `Component`?
pub struct Puzzle {
    pub verts: VertVec<Vec2>,
    pub edges: EdgeVec<(VertIdx, VertIdx)>,
    pub cells: CellVec<Cell>,

    pub square_size: UVec2,
}

/// A `Cell` into which [`Pip`]s can be added
#[derive(Debug, Clone)]
pub struct Cell {
    pub verts: Vec<VertIdx>, // Going clockwise
    pub neighbours: Vec<(EdgeIdx, CellIdx)>,
    pub pips: Pips,
}

/// A (possibly incorrect) attempt to solve a [`Puzzle`]
#[derive(Debug, Clone)]
pub struct Solution {
    pub line: HashSet<EdgeIdx>,
    pub regions: Vec<(Vec<CellIdx>, Pips)>,
    pub pip_count: Option<Pips>,
}

impl Puzzle {
    /// Load a puzzle from a string.  The index of the first '|' indicates the width, and every
    /// other character represents a cell.  All lines are expected to be delimited by '|'s.
    ///
    /// For example, `" 2 |1 1|212"` generates a grid like:
    /// ```text
    /// +   +   +   +
    ///       2
    /// +   +   +   +
    ///   1       1
    /// +   +   +   +
    ///   2   1   2
    /// +   +   +   +
    /// ```
    // TODO: Don't panic on invalid input
    pub fn from_single_line(line: &str) -> Self {
        let rows = line.split("|").collect_vec();
        let width = rows[0].len();
        let height = rows.len();
        assert!(rows.iter().all(|r| r.len() == width)); // Check all rows have same lengths

        // Read pip counts
        let mut pips_in_each_cell = Vec::new();
        for row in rows {
            for cell_char in row.chars() {
                match cell_char {
                    '0'..='9' => pips_in_each_cell.push(Pips(cell_char as usize - '0' as usize)),
                    ' ' => pips_in_each_cell.push(Pips(0)),
                    c => panic!("Unknown char '{c}'"),
                };
            }
        }
        assert_eq!(pips_in_each_cell.len(), width * height);

        Self::rectangular(width, height, pips_in_each_cell)
    }

    /// Make a new `Puzzle` of a rectangular grid of the given size
    fn rectangular(width: usize, height: usize, pips: impl IntoIterator<Item = Pips>) -> Self {
        let mut pip_iter = pips.into_iter();

        // Converters for coords => indices
        let get_vert_idx = |x: usize, y: usize| VertIdx::new(y * (width + 1) + x);
        let cell_idx = |x: usize, y: usize| CellIdx::new(y * width + x);

        // Create vertices
        let mut verts = VertVec::new();
        for y in 0..height + 1 {
            for x in 0..width + 1 {
                verts.push(Vec2::new(x as f32, (height - y) as f32));
            }
        }
        // Create edges
        let mut edges = EdgeVec::new();
        for y in 0..height {
            for x in 0..width + 1 {
                edges.push((get_vert_idx(x, y), get_vert_idx(x, y + 1))); // Vertical
            }
        }
        for y in 0..height + 1 {
            for x in 0..width {
                edges.push((get_vert_idx(x, y), get_vert_idx(x + 1, y))); // Horizontal
            }
        }
        let get_edge_idx_between = |v1: VertIdx, v2: VertIdx| {
            EdgeIdx::new(
                edges
                    .iter()
                    .position(|verts| *verts == (v1, v2) || *verts == (v2, v1))
                    .unwrap(),
            )
        };
        // Create cells
        let mut cells = CellVec::new();
        for y in 0..height {
            for x in 0..width {
                let tl = get_vert_idx(x, y);
                let tr = get_vert_idx(x + 1, y);
                let br = get_vert_idx(x + 1, y + 1);
                let bl = get_vert_idx(x, y + 1);
                // Compute cell neighbours
                let mut neighbours = Vec::new();
                for (v1, v2, dx, dy) in [
                    (tl, tr, 0, -1),
                    (tr, br, 1, 0),
                    (br, bl, 0, 1),
                    (bl, tl, -1, 0),
                ] {
                    // Test if the cell over this edge is actually in the puzzle
                    let neighbour_x = x as isize + dx;
                    let neighbour_y = y as isize + dy;
                    if (0..width as isize).contains(&neighbour_x)
                        && (0..height as isize).contains(&neighbour_y)
                    {
                        neighbours.push((
                            get_edge_idx_between(v1, v2),
                            cell_idx(neighbour_x as usize, neighbour_y as usize),
                        ));
                    }
                }
                // Build cell
                cells.push(Cell {
                    verts: vec![tl, tr, br, bl],
                    neighbours,
                    pips: pip_iter.next().unwrap(),
                });
            }
        }

        Puzzle {
            verts,
            edges,
            cells,

            square_size: UVec2::new(width as u32, height as u32),
        }
    }

    /// Get the [`Solution`] defined by the given `line`.
    pub fn solution(&self, line: HashSet<EdgeIdx>) -> Solution {
        let regions = self.regions(&line);
        let pip_count = self.solution_number(&regions);
        Solution {
            line,
            regions,
            pip_count,
        }
    }

    /// Divide the cells into regions who's borders are the given `line`.
    fn regions(&self, line: &HashSet<EdgeIdx>) -> Vec<(Vec<CellIdx>, Pips)> {
        let mut regions = Vec::new();

        let mut regionless_cells: HashSet<CellIdx> =
            (0..self.cells.len()).map(CellIdx::new).collect();
        // Repeatedly find cells that aren't already in a region ...
        while let Some(region_starting_cell) = regionless_cells.iter().next().copied() {
            let mut cells_in_region = Vec::new();
            let mut total_pips = Pips::ZERO;
            // ... and run BFS starting from the `region_starting_cell`
            let mut frontier = vec![region_starting_cell];
            while let Some(cell_idx) = frontier.pop() {
                let is_cell_new = regionless_cells.remove(&cell_idx);
                if is_cell_new {
                    // Add cell to this region
                    cells_in_region.push(cell_idx);
                    total_pips += self.cells[cell_idx].pips;
                    // Add this cell's neighbours to the frontier
                    for (edge, neighbour) in &self.cells[cell_idx].neighbours {
                        if !line.contains(edge) {
                            frontier.push(*neighbour);
                        }
                    }
                }
            }
            regions.push((cells_in_region, total_pips));
        }
        regions
    }

    /// If the puzzle is correctly solved, return `Some(p)` so that every region contains either
    /// 0 or `p` pips.  If the puzzle isn't correctly solved, return `None`.
    fn solution_number(&self, regions: &[(Vec<CellIdx>, Pips)]) -> Option<Pips> {
        let mut non_zero_pip_counts = regions
            .iter()
            .map(|(_, pips)| *pips)
            .filter(|pips| *pips > Pips::ZERO);
        let first_count = non_zero_pip_counts.next().unwrap_or(Pips::ZERO);
        for pip_count in non_zero_pip_counts {
            if pip_count != first_count {
                return None; // Some region had a different pip count to the others
            }
        }
        Some(first_count) // If all non-zero regions have the same count, the puzzle is solved
    }

    /// Prints a picture of the given `line` running through this `Puzzle`.
    // TODO: Make this not upside-down
    pub fn print_line(&self, line: &HashSet<EdgeIdx>) {
        let width = self.square_size.x as usize;
        let height = self.square_size.y as usize;
        let num_horizontal_edges = width * (height + 1);
        let num_vertical_edges = (width + 1) * height;

        let mut pips_iter = self.cells.iter().map(|cell| cell.pips.0);
        let mut vertical_edges = (0..num_vertical_edges).map(EdgeIdx::new);
        let mut horizontal_edges =
            (0..num_horizontal_edges).map(|i| EdgeIdx::new(i + num_vertical_edges));

        for y in 0..height + 1 {
            // Print line of horizontal edges
            for _ in 0..width {
                print!("+");
                match line.contains(&horizontal_edges.next().unwrap()) {
                    true => print!("---"),
                    false => print!("   "),
                }
            }
            println!("+");
            // Print line of vertical edges and each cell's pip count
            let mut next_vertical_edge_char = || -> char {
                match line.contains(&vertical_edges.next().unwrap()) {
                    true => '|',
                    false => ' ',
                }
            };
            if y < height {
                for _ in 0..width {
                    print!("{} ", next_vertical_edge_char());
                    match pips_iter.next().unwrap() {
                        0 => print!("  "),
                        num_pips => print!("{num_pips} "),
                    }
                }
                println!("{}", next_vertical_edge_char());
            }
        }
    }
}

/// Number of 'dice' pips in some region (usually a cell)
#[derive(Debug, Clone, Copy, PartialOrd, Ord, PartialEq, Eq, Hash)]
pub struct Pips(pub usize);

impl Pips {
    pub const ZERO: Self = Pips(0);
}

impl Add for Pips {
    type Output = Self;

    fn add(self, rhs: Self) -> Self::Output {
        Self(self.0 + rhs.0)
    }
}

impl AddAssign for Pips {
    fn add_assign(&mut self, rhs: Self) {
        self.0 += rhs.0;
    }
}

impl Sum for Pips {
    fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
        let mut total = Self::ZERO;
        for p in iter {
            total += p;
        }
        total
    }
}

index_vec::define_index_type! { pub struct VertIdx = usize; }
index_vec::define_index_type! { pub struct EdgeIdx = usize; }
index_vec::define_index_type! { pub struct CellIdx = usize; }
pub type VertVec<T> = index_vec::IndexVec<VertIdx, T>;
pub type EdgeVec<T> = index_vec::IndexVec<EdgeIdx, T>;
pub type CellVec<T> = index_vec::IndexVec<CellIdx, T>;

File: src/puzzle.ts
/* Abstract representations of puzzles */

/// Abstract representation of a `Puzzle`, without any attached lines or solution
class Puzzle {
  verts: Vec2[];
  edges: { v1: number; v2: number }[];
  cells: Cell[];

  total_num_pips: number;
  solutions: number[];
  grid_width: number;
  grid_height: number;

  constructor(pattern: string, solutions: number[]) {
    /* Values used by the rest of the game */
    this.solutions = solutions;

    // Parse string into a list of pips in each cell
    let pip_lines = pattern.split("|");
    let width = pip_lines[0].length;
    let height = pip_lines.length;

    // Create vertices
    this.verts = [];
    for (let y = 0; y < height + 1; y++) {
      for (let x = 0; x < width + 1; x++) {
        this.verts.push(new Vec2(x, y));
      }
    }
    let vert_idx = (x: number, y: number) => y * (width + 1) + x;

    this.edges = [];
    // Vertical edges
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width + 1; x++) {
        this.edges.push({ v1: vert_idx(x, y), v2: vert_idx(x, y + 1) });
      }
    }
    // Horizontal edges
    for (let y = 0; y < height + 1; y++) {
      for (let x = 0; x < width; x++) {
        this.edges.push({ v1: vert_idx(x + 1, y), v2: vert_idx(x, y) });
      }
    }

    // Cells
    this.cells = [];
    this.total_num_pips = 0;
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        // Get the vertices surrounding this cell
        const tl = vert_idx(x, y);
        const tr = vert_idx(x + 1, y);
        const br = vert_idx(x + 1, y + 1);
        const bl = vert_idx(x, y + 1);
        // Get the cells which neighbour this one as well as the corresponding edges
        const possible_neighbours = [
          { v1: tl, v2: tr, dx: 0, dy: -1 }, // top
          { v1: tr, v2: br, dx: 1, dy: 0 }, // right
          { v1: br, v2: bl, dx: 0, dy: 1 }, // bottom
          { v1: bl, v2: tl, dx: -1, dy: 0 }, // left
        ];
        const neighbours = [];
        for (const { v1, v2, dx, dy } of possible_neighbours) {
          let new_x = x + dx;
          let new_y = y + dy;
          // Neighbour is only valid if the opposite cell is actually in the grid
          if (new_x < 0 || new_x >= width || new_y < 0 || new_y >= height) continue;
          // Add neighbour
          let cell_idx = new_y * width + new_x;
          const edge_idx = this.connecting_edge(v1, v2)!;
          neighbours.push({ edge_idx, cell_idx });
        }

        let num_pips = parseInt(pip_lines[y][x]) || 0;
        this.total_num_pips += num_pips;
        this.cells.push({
          verts: [tl, tr, br, bl],
          centre: new Vec2(x + 0.5, y + 0.5),
          num_pips,
          neighbours,
        });
      }
    }

    // Check that the solutions look vaguely sane
    let min_pip_count = Infinity;
    let total_pip_count = 0;
    for (const { num_pips } of this.cells) {
      min_pip_count = Math.min(min_pip_count, num_pips);
      total_pip_count += num_pips;
    }
    for (const s of this.solutions) {
      if (total_pip_count % s !== 0) {
        console.warn(
          `Solution ${s} of "${pattern}" doesn't divide the total of ${total_pip_count}`,
        );
      }
      if (s < min_pip_count) {
        console.warn(`Solution ${s} of "${pattern}" is smaller than the min pip count`);
      }
    }

    // Find the bbox of the puzzle
    let min = new Vec2(Infinity, Infinity);
    let max = new Vec2(-Infinity, -Infinity);
    for (const v of this.verts) {
      min = Vec2.min(min, v);
      max = Vec2.max(max, v);
    }
    let rect_centre = Vec2.lerp(min, max, 0.5);

    // Translate puzzle so the centre of the bounding box is at (0, 0)
    for (let v = 0; v < this.verts.length; v++) {
      this.verts[v] = this.verts[v].sub(rect_centre);
    }
    for (const c of this.cells) {
      c.centre = c.centre.sub(rect_centre);
    }
    // Store width and height of this puzzle's bounding box
    this.grid_width = max.x - min.x;
    this.grid_height = max.y - min.y;
  }

  get_solution(line: number[]): Solution {
    console.assert(line[0] === line[line.length - 1]); // Check that line forms a loop
    // Determine which edges are in the line
    let is_edge_in_line = [];
    for (const _ of this.edges) is_edge_in_line.push(false);
    for (let i = 0; i < line.length - 1; i++) {
      const edge_idx = this.connecting_edge(line[i], line[i + 1])!;
      is_edge_in_line[edge_idx] = true;
    }

    // Use the line to split the cells into regions (i.e. strongly connected components of the
    // (dual of the) cell graph where edges under the line are removed)
    let is_cell_in_region = [];
    for (const _ of this.cells) is_cell_in_region.push(false);
    let regions = [];
    while (true) {
      // Find a cell which isn't yet in a region
      let region_start_cell = undefined;
      for (let i = 0; i < this.cells.length; i++) {
        if (!is_cell_in_region[i]) {
          region_start_cell = i;
          break;
        }
      }
      // If no such cell exists, all cells are in regions and we're done
      if (region_start_cell === undefined) break;
      // Take this new cell and explore its entire region using DFS
      let cells_in_region = [];
      let num_pips_in_region = 0;
      let frontier = [region_start_cell];
      while (frontier.length > 0) {
        let next_cell_idx = frontier.pop()!;
        let next_cell = this.cells[next_cell_idx];
        if (is_cell_in_region[next_cell_idx]) continue; // Cell has already been explored
        // Add cell to this region
        is_cell_in_region[next_cell_idx] = true;
        cells_in_region.push(next_cell_idx);
        num_pips_in_region += next_cell.num_pips;
        // Add cell's neighbours to the frontier
        for (const { edge_idx, cell_idx } of next_cell.neighbours) {
          if (is_edge_in_line[edge_idx]) continue; // Can't connect region over line
          frontier.push(cell_idx);
        }
      }
      // Once BFS has fully explored the region, add this region as complete
      regions.push({ num_pips: num_pips_in_region, cells: cells_in_region });
    }

    // Use the pip counts to check if the regions actually make a valid solution
    const pip_counts = regions.map((r) => r.num_pips).filter((pips) => pips > 0);
    const pip_group_size = pip_counts[0];
    const is_correct = pip_counts.length > 1 && pip_counts.every((p) => p == pip_group_size);

    // Package the solution and return
    return { is_correct, pip_group_size, regions };
  }

  // Find the nearest point to `(p_x, p_y)` on any edge in the puzzle, *excluding* those already on
  // the given `line`.
  nearest_edge_point_extending_line(p: Vec2, line: number[]): NearestEdge {
    let nearest = undefined;
    for (let edge_idx = 0; edge_idx < this.edges.length; edge_idx++) {
      let { v1, v2 } = this.edges[edge_idx];

      // Test if this edge is extending the line but not already on it
      let is_extending_line = line.length === 0 ||
        v1 === line[line.length - 1] ||
        v2 === line[line.length - 1];
      let is_on_line = false;
      for (let i = 0; i < line.length - 2; i++) {
        if (
          (line[i] === v1 && line[i + 1] === v2) ||
          (line[i] === v2 && line[i + 1] === v1)
        ) {
          is_on_line = true;
          break;
        }
      }
      if (is_on_line || !is_extending_line) {
        continue;
      }

      // If this edge is a valid extension/contraction of the line, then it is a candidate for
      // being the closest edge to the point.

      // Find nearest point on line
      let s = this.verts[v1];
      let d = this.verts[v2].sub(s);
      let d_dot_p_minus_s = d.dot(p.sub(s));
      let lambda = d_dot_p_minus_s / d.square_length();
      lambda = Math.max(0, Math.min(1, lambda)); // Clamp to the line
      let nearest_point = s.add(d.mul(lambda));
      // Compute distance
      let distance = p.sub(nearest_point).length();
      if (nearest === undefined || distance < nearest.distance) {
        nearest = {
          edge_idx,
          lambda,
          point: nearest_point,
          distance,
        };
      }
    }
    return nearest!;
  }

  nearest_vertex(point: Vec2): NearestVert {
    let nearest_vert = undefined;

    for (let vert_idx = 0; vert_idx < this.verts.length; vert_idx++) {
      let { x: vert_x, y: vert_y } = this.verts[vert_idx];
      let dX = point.x - vert_x;
      let dY = point.y - vert_y;
      let distance = Math.sqrt(dX * dX + dY * dY);
      // ... if this is the new closest vertex, update the interaction
      if (nearest_vert === undefined || distance < nearest_vert.distance) {
        nearest_vert = { point, vert_idx, distance };
      }
    }

    return nearest_vert!; // This is only 'undefined' if puzzle has no vertices
  }

  connecting_edge(vert_1: number, vert_2: number): number | undefined {
    for (let i = 0; i < this.edges.length; i++) {
      const { v1, v2 } = this.edges[i];
      if (vert_1 === v1 && vert_2 === v2) return i;
      if (vert_1 === v2 && vert_2 === v1) return i;
    }
    return undefined;
  }
}

type Cell = {
  verts: number[];
  centre: Vec2;
  num_pips: number;
  neighbours: { edge_idx: number; cell_idx: number }[];
};

type Solution = {
  is_correct: boolean;
  pip_group_size: number;
  regions: Region[];
};

type Region = { num_pips: number; cells: number[] };

type NearestEdge = {
  edge_idx: number;
  lambda: number;
  point: Vec2;
  distance: number;
};

type NearestVert = {
  point: Vec2;
  vert_idx: number;
  distance: number;
};
